File: neetcode/problems/leetcode70/solution.js
const cached = () => {
	const cache = {};

	const climbStairs = n => {
		if (n <= 3) return n;

		if (!cache[n - 1]) cache[n - 1] = climbStairs(n - 1);
		if (!cache[n - 2]) cache[n - 2] = climbStairs(n - 2);

		return cache[n - 1] + cache[n - 2];
	};

	return climbStairs;
};

const climbStairsRec = cached();

const climbStairs = n => {
	if (n <= 3) return n;

	let a = 1;
	let b = 2;
	let c = 3;

	for (let i = 4; i <= n; i++) {
		a = b;
		b = c;
		c = a + b;
	}

	return c;
};

File: neetcode/problems/leetcode70/solution.py
class SolutionRec:
    def climbStairs(self, n: int) -> int:
        cache = {}

        def climb(n: int) -> int:
            if n <= 3:
                return n

            if n not in cache:
                cache[n] = climb(n - 1) + climb(n - 2)

            return cache[n]

        return climb(n)


class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 3:
            return n

        a = 1
        b = 2
        c = 3

        for _ in range(3, n):
            a = b
            b = c
            c = a + b

        return c


class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 3:
            return n

        a, b = 1, 2

        for _ in range(3, n + 1):
            a, b = b, a + b

        return b
